torch_geometric/transforms/node_property_split.py

```diff
@@ -0,0 +1,236 @@
+    node property by adding :obj:`split_masks` to the
+    :class:`~torch_geometric.data.Data` object, as proposed in the
+    `Evaluating Robustness and Uncertainty of Graph Models Under Structural
+    Distributional Shifts <https://arxiv.org/abs/2302.13875v1>`__ paper
+    (functional name: :obj:`node_property_split`)
+
+    It splits the nodes in a given graph into 5 non-intersecting parts
+    based on their structural properties. This can be used for transductive
+    node prediction task with distributional shifts. It considers the
+    in-distribution (ID) and out-of-distribution (OOD) subsets of nodes.
+    The ID subset includes training, validation and testing parts, while
+    the OOD subset includes validation and testing parts. As a result, it
+    creates 5 associated node mask arrays for each graph, 3 of which are
+    for the ID nodes: :obj:`"in_train_mask"`, :obj:`"in_valid_mask"`,
+    :obj:`"in_test_mask"`, and the remaining 2 — for the OOD nodes:
+    :obj:`"out_valid_mask"`, :obj:`"out_test_mask"`.
+
+    This class implements 3 particular strategies for inducing
+    distributional shifts in graph — based on
+    **popularity**, **locality** or **density**.
+
+    Args:
+        property_name (str): The name of the node property to be used,
+            which must be ``'popularity'``, ``'locality'`` or ``'density'``.
+
+        part_ratios (list): A list of 5 ratio values for training,
+            ID validation, ID test, OOD validation and OOD test parts.
+            The values must sum to 1.0.
+
+        ascending (bool, optional): Whether to sort nodes in the ascending
+            order of the node property, so that nodes with greater values
+            of the property are considered to be OOD (default: :obj:`True`)
+
+        random_seed (int, optional): Random seed to fix for the initial
+            permutation of nodes. It is used to create a random order for
+            the nodes that have the same property values or belong to the
+            ID subset. (default: :obj:`None`)
+
+    Example:
+
+    .. code-block::
+
+        from torch_geometric.transforms import NodePropertySplit
+        from torch_geometric.datasets import graph_generator
+
+        num_nodes = 1000
+        edge_prob = 0.4
+        data = graph_generator.ERGraph(num_nodes, edge_prob)()
+
+        property_name = "popularity"
+        part_ratios = [0.3, 0.1, 0.1, 0.3, 0.2]
+        tranaform = NodePropertySplit(property_name, part_ratios)
+
+        data = transform(data)
+    """
+    def __init__(
+        self,
+        property_name: str,
+        part_ratios: list,
+        ascending: Optional[bool] = True,
+        random_seed: Optional[int] = None,
+    ):
+        assert property_name in [
+            "popularity",
+            "locality",
+            "density",
+        ], "`property_name` has to be 'popularity', 'locality', or 'density'"
+
+        assert sum(part_ratios) == 1.0, "`part_ratios` must sum to 1.0"
```

Comment:
Add an assert statement to verify that the length of `part_ratios` is equal to 5.